### ISSUE TITLE
Gemma 4 12B QAT: 106/600 first-token mismatches with pinned llama.cpp, identical tokenization

### ISSUE THREAD

**Issue by avifenesh** [avifenesh/memra#398]
Gemma 4 12B QAT has a reproducible first-token parity gap against a pinned llama.cpp HTTP oracle: 106 mismatches on 600 byte-identical five-shot Global-MMLU prompts. All 600 prompt token sequences match exactly, including BOS ID 2. The Memra replay matches the earlier Memra run 600/600, and its diagnostic `prime_cache` CLI matches the HTTP top-one text 600/600.

This is a numerical fidelity investigation, not a claim that Memra caused the low benchmark accuracy. Both runtimes score similarly under this raw-prompt, one-token protocol. The exact responsible kernel is unresolved; no fix or default change is included.

| Language, n=300 | Memra correct | llama.cpp correct | Memra unparsed | llama.cpp unparsed | Exact token agreement |
|---|---:|---:|---:|---:|---:|
| English | 154 (51.33%) | 157 (52.33%) | 25 | 24 | 247/300 |
| Hebrew | 99 (33.00%) | 94 (31.33%) | 34 | 32 | 247/300 |

The fixed subset is proportional stratification over prior correct/wrong/unparsed outcomes: English 154/121/25, Hebrew 99/167/34. Selection uses SHA256 ordering with seed `gemma4-diag-20260909`; full manifest SHA256 is `95489aa0276d2ea4e10061dd4992a64fb0631de943600bb997ccfa2a078a77b4`. These are diagnostic subset scores, not full MMLU or unquantized-model scores.

Pins:

- Memra v0.136.0, engine commit `55d83a7cf3948d6d173d28e3e1fd18008fdeb9bf`; server binary SHA256 `61a73b96bb631cdae7ca42816834872e35156e7e5c105ce1bdaace7f91912b33`.
- `google/gemma-4-12b-it-qat-q4_0-gguf`, revision `29d097773436b69ff9feafd636ab4cf873786537`; GGUF SHA256 `93567e57a8fe10b23569b9d9ec38cd005deedf71e29477c421a4b83f418a538b`.
- llama.cpp `f3f1a8f2760f28325a5ec20c05b171e5b7c83a29`, CUDA 13.0.2, RTX 5090 32 GB, all 49 layers offloaded.
- `CohereLabs/Global-MMLU`, revision `0e619dbeb34206cd48705a1a0ea7fb21cae09993`, files `{en,he}/{dev,test}-00000-of-00001.parquet`.

The original and replay Memra boot logs select INT8 MMQ without FP16 mirrors. The artifact has 48 layers, a 1024-token sliding window, zero shared-KV layers and zero per-layer embedding width. A shared-KV implementation gap is not implicated by this artifact. The mismatch occurs on both sides of the window: 91/465 mismatches at <=1024 tokens, 15/135 above it.

| Language / sample ID | Memra ID / text | llama.cpp ID / text | Oracle top-two gap |
|---|---|---|---:|
| en `computer_security/test/33` | 236780 / `"C"` | 565 / `" C"` | 4.414430 |
| en `miscellaneous/test/547` | 236771 / `"0"` | 603 / `" B"` | 4.205450 |
| en `clinical_knowledge/test/52` | 107 / `"\n"` | 565 / `" C"` | 4.068754 |
| en `prehistory/test/145` | 562 / `" A"` | 565 / `" C"` | 3.874479 |
| en `moral_disputes/test/225` | 562 / `" A"` | 565 / `" C"` | 3.798855 |
| en `high_school_psychology/test/438` | 565 / `" C"` | 107 / `"\n"` | 3.696179 |
| he `marketing/test/106` | 603 / `" B"` | 562 / `" A"` | 3.546499 |
| en `professional_psychology/test/456` | 562 / `" A"` | 236776 / `"A"` | 3.497564 |

95 of the 106 token mismatches change the original parser prediction (`text.strip()[:1].upper()`). Median oracle top-two gap among mismatches is 0.460214; the maximum is 4.414430.

Existing-option controls, compared against the same pinned HTTP oracle:

| Arm | Items | Oracle agreement | Default agreement on the same items |
|---|---:|---:|---:|
| FP16 mirrors | 385 | 315 | 319 |
| Non-flash attention | 600 | 497 | 494 |
| FP16 mirrors + non-flash | 32 | 25 | 16 |
| F32 dequantization, GEMM/MMQ disabled, non-flash | 16 | 11 | 8 |

The FP16 arm OOMed on index 385 and is incomplete. The 32-case arm selects 16 largest-margin disagreements under 800 tokens plus 16 agreeing controls. The F32 arm uses the first eight of each group. These controls change outcomes but do not identify a single faulty operation. They are not accuracy estimates.

Oracle integrity and limits:

- A read-only patch immediately before `common_sampler_sample` records top-five values from `llama_get_logits_ith`. Repeating all 600 raw and 600 chat items with that instrumentation reproduced all 1200 outputs exactly. All 1200 recorded argmax IDs equal the reported tokens.
- Standalone llama.cpp direct-API calls are separate controls: they disagree with this HTTP oracle on 80/600 raw items with API defaults and 84/600 after aligning `swa_full` and batch size. The remaining API/server configuration cause is unresolved. They are not substituted into the score table, and the mismatch alone is not proof of an absolute-ground-truth kernel bug.
- Applying the GGUF Gemma Jinja template produces `<|channel>` as the first token on all 600 items. One-token scoring then has 600 unparsed results. This does not measure completed chat-answer quality.

Reproduction commands (separate GPU processes):

```sh
MEMRA_COMPAT=openai MEMRA_MODELS=gemma=/path/model.gguf MEMRA_PORT=8080 PORT=8080 \
MEMRA_NONSTREAM_DEADLINE_GATE=0 MEMRA_TIMEOUT_MS_MAX=1800000 ./memra-server

# Pinned llama.cpp checkout:
cmake -S . -B build -G Ninja -DGGML_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=120 \
  -DLLAMA_CURL=OFF -DLLAMA_BUILD_TESTS=OFF
cmake --build build --target llama-server -j 8
build/bin/llama-server -m /path/model.gguf --host 127.0.0.1 --port 8081 \
  -ngl 99 -c 8192 -b 4096 -ub 2048 -np 1 --jinja --no-warmup
```

Use the pinned parquet files renamed `en-dev.parquet` and `en-test.parquet`. This reproducer checks the exact prompt hash before either request; run each process separately if memory requires it.

<details><summary>Exact clinical_knowledge/test/52 prompt reconstruction and requests</summary>

```python
#!/usr/bin/env python3
import hashlib,json,urllib.request,argparse
import pandas as pd
ap=argparse.ArgumentParser();ap.add_argument('--data',required=True);ap.add_argument('--memra',default='http://127.0.0.1:8080');ap.add_argument('--llama',default='http://127.0.0.1:8081');args=ap.parse_args()
test=pd.read_parquet(args.data+'/en-test.parquet');dev=pd.read_parquet(args.data+'/en-dev.parquet');row=test[test.sample_id=='clinical_knowledge/test/52'].iloc[0]
def block(r,answer):
 s=str(r.question).strip()+'\n'
 for letter,col in zip('ABCD',['option_a','option_b','option_c','option_d']):s+=f'{letter}. {str(r[col]).strip()}\n'
 return s+'Answer:'+(' '+str(r.answer).strip()+'\n\n' if answer else '')
prompt=''.join(block(r,True) for _,r in dev[dev.subject==row.subject].head(5).iterrows())+block(row,False)
assert hashlib.sha256(prompt.encode()).hexdigest()=='cfed1dae01ad0e01d1ebc416180d6dd73fa6a8e67884d9ad56fcf7ff3694bebe'
for name,url,body in [('memra',args.memra+'/v1/completions',{'model':'gemma','prompt':prompt,'temperature':0,'max_tokens':1}),('llama',args.llama+'/completion',{'prompt':prompt,'temperature':0,'n_predict':1,'n_probs':5,'post_sampling_probs':False,'return_tokens':True,'cache_prompt':False,'seed':0})]:
 req=urllib.request.Request(url,data=json.dumps(body).encode(),headers={'Content-Type':'application/json'})
 with urllib.request.urlopen(req,timeout=300) as r:print(name,json.dumps(json.load(r),ensure_ascii=False))

```

</details>

For this case the expected Memra output is newline (ID 107), versus llama.cpp ` C` (ID 565). Prompt SHA256: `cfed1dae01ad0e01d1ebc416180d6dd73fa6a8e67884d9ad56fcf7ff3694bebe`.

<details><summary>Top-five raw logits for that case</summary>

```json
{
  "memra": [
    {
      "id": 107,
      "logit": 21.820911407,
      "text": "\n"
    },
    {
      "id": 236770,
      "logit": 21.723747253,
      "text": "1"
    },
    {
      "id": 236800,
      "logit": 20.707910538,
      "text": "3"
    },
    {
      "id": 236780,
      "logit": 20.456245422,
      "text": "C"
    },
    {
      "id": 236778,
      "logit": 20.164531708,
      "text": "2"
    }
  ],
  "llama_http": [
    {
      "id": 565,
      "logit": 26.506031036376953
    },
    {
      "id": 603,
      "logit": 22.43727684020996
    },
    {
      "id": 622,
      "logit": 21.633258819580078
    },
    {
      "id": 562,
      "logit": 20.428693771362305
    },
    {
      "id": 108,
      "logit": 18.5263671875
    }
  ]
}
```

</details>

Complete local receipts: `/home/avifenesh/evals/hebrew-agentic-base-20260909/gemma4-diag/`. Curated report, 600-row divergence data, manifest, raw oracle logits, instrumentation patch, build hashes and commands: `research/hebrew-agentic-base-20260909/measure/gemma4-diag/` in the private operations repository. The reproduction above is self-contained and uses public pinned inputs.

Follow-up: establish a single matched numerical oracle contract, then localize the remaining logits divergence by layer/operator. Do not infer base-model quality or roll out a serving flag from this diagnostic.
